[24.1] Fix upload when current history changes

davelopez opened the pull request:
Fixes #18662

<del>The upload queue will now be reused if it was not started/running before when you change the target history.</del>

This will ensure that when we change the current history in the middle of the upload each dataset goes to the history where it was scheduled for the first time.

As a bonus, the name of the current history is shown at the top of the upload modal to be more clear about the target history (thanks @ahmedhamidawan for the idea). If this is too much for a fix we can move those commits to dev or drop them if people think is not necessary.

![image](https://github.com/user-attachments/assets/b14eb297-92f5-4275-91a5-1f2bef725d90)


## How to test the changes?
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - Follow the steps in #18662

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Expose currentHistory in useUserHistories composable

davelopez committed:
Display target history name in upload modal

davelopez committed:
Fix initialization logic for upload queue

When the current history changes we need to either "reuse" or "create" a new upload queue

mvdbeek commented on client/src/components/Upload/DefaultBox.vue:
```diff
@@ -99,7 +100,16 @@ const listExtensions = computed(() => props.effectiveExtensions.filter((ext) =>
+watch(
+    () => props.historyId,
+    () => {
+        queue.value = initUploadQueue();
```
Is this not a problem for uploads that are still running and targeting another history ?

mvdbeek replied:
I would have assumed you need to associate each upload with a history

davelopez replied:
We are still associating the whole queue with a history. The only difference is that before we always created a new queue whenever the current history changed, but now we will re-target to the new history only IF the queue was not "started". If it was already running, it will continue running in the old history and we create a "new queue" for the "new history" so whatever we add next is going to be associated with the new one.

So, in summary, once we press "Start" whatever is in the queue will go to the current history if I didn't misunderstand anything.

mvdbeek replied:
But you lost the old queue then, how does the progress bar update ?

davelopez replied:
In my local tests the progress "works". I uploaded 100 files to "history A" and then switched to "history B" in the middle of the upload. The progress keeps going and the files get uploaded to the history A. The only thing is that, if you open the uploader, you need to wait until the progress is completed before you can try to upload anything to "history B". The UI won't let you otherwise.

If this is not the desired behavior I can try to change the implementation to maintain a unique queue and associate the history with each individual item when the user presses "Start" as you suggest.

mvdbeek replied:
OK, that's fine for 24.1. Yes, I think you should be able to start a new upload at any point, but if that's a larger change targeting dev is fine.

davelopez replied:
I will keep exploring a bit the option to associate the history with individual items and see if I can make it work without too many changes, if not, we can go with these changes as a fix :+1:

davelopez replied:
Actually, this is a cleaner and simpler solution! Thanks for suggesting it!

It now lets you pause and schedule upload datasets to the new history and continue. Better user experience I think.

![UploadSwithHistoryFix](https://github.com/user-attachments/assets/6f9c1bbd-49a0-4a63-9713-3da4953ec7d2)

ahmedhamidawan replied:
It might just be me but isn't it a little difficult to decipher the difference between what will go to `Old History` and what to `New History`? As in, would it be nicer to have an `<hr>` between the file queues in the uploader?

davelopez replied:
This is a good point and we should definitely consider how to communicate this better in the future uploader redesign or in dev :+1: 
That said, I would prefer to keep the fix as minimal as possible, I already stretched it a bit by adding the history name at the top :stuck_out_tongue_closed_eyes:

ahmedhamidawan replied:
Fair point, we can keep the change minimal for now

ElectronicBlueberry commented on client/src/components/Upload/DefaultBox.vue:
```diff
@@ -99,7 +100,16 @@ const listExtensions = computed(() => props.effectiveExtensions.filter((ext) =>
-const queue = computed(() => createUploadQueue());
+onMounted(() => {
+    queue.value = initUploadQueue();
+});
```
To avoid duplication, you can put `{ immediate: true }` in the watcher options

davelopez replied:
I keep forgetting that, thanks!

ElectronicBlueberry replied:
There's also this neat [shorthand](https://vueuse.org/shared/watchImmediate/)

davelopez committed:
Use watchImmediate instead of onMounted and watch

Co-authored-by: Laila Los <44241786+ElectronicBlueberry@users.noreply.github.com>

davelopez committed:
Associate target history with each upload item when upload starts

This will ensure that when we change the current history in the middle of the upload each dataset goes where it was scheduled for the first time.

davelopez committed:
Increase default lazy load value to 150 in DefaultBox.vue

davelopez committed:
Submit remote files as single batch request per target history

This is unlikely to happen, but it will be consistent if we switch around histories and add remote files to different ones while pausing and continuing upload progress on them.

davelopez committed:
Adapt upload-queue tests
